hw/chisel/doc/xdma.md

```diff
@@ -0,0 +1,238 @@
+
+### Extendable Datapath Extension to Enable *Computation on Copy* Capability
+
+The Datapath Extension is a critical component allowing on-the-fly data processing such as matrix transposing, max-pooling, average-pooling, and data stream compression/decompression. By supplying the necessary parameters to the XDMA generator, extension modules are instantiated sequentially, with each module's output feeding into the next module's input.
+
+Designing a custom datapath extension is straightforward thanks to Chisel's flexibility. The pipeline is automatically inserted, CSR configs are automatically wired, infrastructure for enabling/disabling extensions is added, and the corresponding C header for runtime is automatically generated. Integrating native extensions written in Chisel or extensions written in SystemVerilog is remarkably simple.
+
+### Distributed Datapath for Cluster-to-Cluster or Host-to-Cluster Memory Access
+
+XDMA can perform data copying within the same memory region or across different regions. If the source and destination addresses are from the same region, data is looped back directly without utilizing the AXI, conserving energy. If addresses span different regions, the data is handled in three steps:
+1. The controller recognizes that the source data is from a non-local region and flattens the configuration into a one-dimensional frame.
+2. The configuration frame is sent to the remote XDMA to request the data.
+3. The data is relayed back from the remote to the local site.
+
+An illustration of the XDMA call chain is shown below:
+
+![XDMA Call Chain](xdma_figures/xdma_callchain.png)
+
+Note: The AXI protocol adaptor is still under development and the architecture may change.
+
+## CSR Address Definition
+
+Consider the following configuration:
+
+```hjson
+// Key configuration items only:
+
+  // CFG items inside cluster
+  cluster: {
+    addr_width: 48,
+    data_width: 64,
+    dma_data_width: 512
+  },
+  // CFG items inside XDMA core template
+  xdma_core_template: {
+     snax_xdma_cfg: {
+         reader_buffer: 8, 
+         writer_buffer: 8, 
+         reader_agu_dimension: 3, 
+         writer_agu_dimension: 3, 
+         HasTransposer: 3, 
+         HasMemset: 1, 
+         HasMaxPool: 2
+     }
+  }
+```
+
+This configuration specifies eight channels for both reader and writer TCDM ports, employing three-dimensional address generation. The innermost dimension is used explicitly by the eight parallel channels, and its value can be reduced to disable some channels. The other two dimensions are managed by the Address Generation Unit (AGU).
+
+The enabled extensions — **HasTransposer**, **HasMemset**, and **HasMaxPool** — add bypass signals and custom CSRs to the configuration list. The positive number at the back means it will be enabled, and the size of the number indicate its sequence - the one with smaller value will in the front of one with larger value. 
```

Comment:
this is cool!

Reply:
Haha the effort to make the "config file" the only source of truth is tremendous. Imagine the case that the users add the new extension. If after writing the actual implementation, the users still need to modify generator to do the correct generation of the new extension. It is not elegant at all, and this asks the users to understand scala before they can do integration. To address this, I use meta programming provided by Scala (and Java), which means while Chisel generates SystemVerilog, Chisel also generates Chisel itself. See #207 for more information.